unblob/pool.py

```diff
@@ -0,0 +1,117 @@
+import abc
+import multiprocessing as mp
+import sys
+from multiprocessing.queues import JoinableQueue
+from typing import Any, Callable, Union
+
+from .logging import multiprocessing_breakpoint
+
+
+class PoolBase(abc.ABC):
+    @abc.abstractmethod
+    def submit(self, args):
+        pass
+
+    @abc.abstractmethod
+    def process_until_done(self):
+        pass
+
+    def start(self):
+        pass
+
+    def close(self):
+        pass
+
+    def __enter__(self):
+        self.start()
+        return self
+
+    def __exit__(self, *args):
+        self.close()
+
+
+class Queue(JoinableQueue):
+    def is_empty(self) -> bool:
+        """Checks if all ``task_done`` has been called for all items.
+        Based on ``multiprocessing.JoinableQueue.join``."""
+        with self._cond:  # type: ignore
+            return self._unfinished_tasks._semlock._is_zero()  # type: ignore
+
+
+def _worker_process(handler, input, output):
+    sys.breakpointhook = multiprocessing_breakpoint
+    while True:
+        args = input.get()
+        result = handler(args)
+        output.put(result)
+
+
+class MultiPool(PoolBase):
+    def __init__(
+        self,
+        process_num: int,
+        handler: Callable[[Any], Any],
+        *,
+        result_callback: Callable[["MultiPool", Any], Any]
+    ):
+        if process_num <= 0:
+            raise ValueError("At process_num must be greater than 0!")
+
+        if process_num == 1:
+            raise ValueError("Use SinglePool instead of MultiPool for 1 process!")
+
+        self._result_callback = result_callback
+        self._input = Queue(ctx=mp.get_context())
+        self._output = mp.Queue()
+        self._procs = [
+            mp.Process(
+                target=_worker_process,
+                args=(handler, self._input, self._output),
+            )
+            for _ in range(process_num)
+        ]
+
+    def start(self):
+        for p in self._procs:
+            p.start()
+
+    def close(self):
+        self._input.join()
+        for p in self._procs:
+            p.terminate()
+            p.join()
+
+    def submit(self, args):
+        self._input.put(args)
+
+    def process_until_done(self):
+        while not self._input.is_empty():
+            result = self._output.get()
+            self._result_callback(self, result)
+            self._input.task_done()
+
+
+class SinglePool(PoolBase):
+    def __init__(self, handler, *, result_callback):
+        self._handler = handler
+        self._result_callback = result_callback
+        self._args = []
+
+    def submit(self, args):
+        self._args.append(args)
+
+    def process_until_done(self):
+        for args in self._args:
```

Comment:
This is incorrect, as the callback may call submit, changing the _args list while iterating over it via the for loop. Can be changed to a while loop and using pop though to retrieve the next element

Reply:
Really? 
* The `handler` foes not have the `pool` itself, only the `result_callback`.
* In the tests, the `result_callback` is already submitting to the pool in the test, and it is passed.

What can't I see here?

Reply:
the callback is called from line 106 inside the for loop which iterates over the same array. I wouldn't count on this working every time on each supported python version and arbitrary number of appends.